### PR TITLE
feat(repo_blob): add CatFileBlob

### PR DESCRIPTION
**Pull request by jonatan5524**
### Describe the pull request

Add a method that returns any Blob object by its corresponding hash.

Link to the issue: gogs/gogs#5979

### Checklist

- [x] I agree to follow the [Code of Conduct](https://go.dev/conduct) by submitting this pull request.
- [x] I have read and acknowledge the [Contributing guide](https://github.com/gogs/gogs/blob/main/.github/CONTRIBUTING.md).
- [x] I have added test cases to cover the new code.

### CODE REVIEW

File: repo_blob.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+	if typ != ObjectBlob {
+		return nil, ErrNotBlob
+	}
+
+	return &Blob{
+		TreeEntry: &TreeEntry{
+			mode: EntryBlob,
+			typ:  ObjectBlob,
+			id:   MustIDFromString(rev),
```
Comment (unknwon) [inline repo_blob.go:22]: There is no guarantee the `rev` being passed here is always a full-length SHA, I think we need to same thing as in `CatFileCommit`:

https://github.com/gogs/git-module/blob/a2567c624d4263a01df4a5405ebe065a9f96e0c3/repo_commit.go#L95-L98

**Review by unknwon — CHANGES_REQUESTED** [review]
Thanks for the PR! Just one problem.

File: repo_blob_test.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+func TestRepository_CatFileBlob(t *testing.T) {
+	t.Run("not a blob", func(t *testing.T) {
+		_, err := testrepo.CatFileBlob("007cb92318c7bd3b56908ea8c2e54370245562f8")
+		assert.Equal(t, ErrNotBlob, err)
+	})
+
+	t.Run("get a blob", func(t *testing.T) {
+		b, err := testrepo.CatFileBlob("021a721a61a1de65865542c405796d1eb985f784")
+		require.NoError(t, err)
+
```
Comment (unknwon) [inline repo_blob_test.go:22]:
```suggestion
```

File: repo_blob_test.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
+// Copyright 2022 The Gogs Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package git
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
```
Comment (unknwon) [inline repo_blob_test.go:10]:
```suggestion
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
```

File: repo_blob.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2022 The Gogs Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package git
+
+// CatFileBlob returns the blob corresponding to the given revision of the repository.
+func (repo *Repository) CatFileBlob(rev string) (*Blob, error) {
+	revHash, err := r.RevParse(rev, RevParseOptions{Timeout: opt.Timeout}) //nolint
```
Comment (unknwon) [inline repo_blob.go:9]:
```suggestion
	rev, err := r.RevParse(rev, RevParseOptions{Timeout: opt.Timeout}) //nolint
```

You can reuse the same variable.

File: repo_blob_test.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+func TestRepository_CatFileBlob(t *testing.T) {
+	t.Run("not a blob", func(t *testing.T) {
+		_, err := testrepo.CatFileBlob("007cb92318c7bd3b56908ea8c2e54370245562f8")
+		assert.Equal(t, ErrNotBlob, err)
+	})
+
+	t.Run("get a blob", func(t *testing.T) {
+		b, err := testrepo.CatFileBlob("021a721a61a1de65865542c405796d1eb985f784")
```
Comment (unknwon) [inline repo_blob_test.go:21]: Could you also add a test case like `testrepo.CatFileBlob("021a721")`?

**Comment by unknwon** [comment]
Thanks for the followup!

**Review by unknwon — APPROVED** [review]

File: repo_blob.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+package git
+
+// CatFileBlob returns the blob corresponding to the given revision of the repository.
+func (repo *Repository) CatFileBlob(rev string) (*Blob, error) {
+	rev, err := r.RevParse(rev, RevParseOptions{Timeout: opt.Timeout}) //nolint
+	if err != nil {
+		return nil, err
+	}
+
+	typ, err := repo.CatFileType(revHash)
```
Comment (unknwon) [inline repo_blob.go:14]:
```suggestion
	typ, err := repo.CatFileType(rev)
```

File: repo_blob.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2022 The Gogs Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package git
+
+// CatFileBlob returns the blob corresponding to the given revision of the repository.
+func (repo *Repository) CatFileBlob(rev string) (*Blob, error) {
```
Comment (unknwon) [inline repo_blob.go:8]:
```suggestion
func (r *Repository) CatFileBlob(rev string) (*Blob, error) {
```

File: repo_blob.go [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+	rev, err := r.RevParse(rev, RevParseOptions{Timeout: opt.Timeout}) //nolint
+	if err != nil {
+		return nil, err
+	}
+
+	typ, err := repo.CatFileType(rev)
```
Comment (unknwon) [inline repo_blob.go:14]:
```suggestion
	typ, err := r.CatFileType(rev)
```

File: repo_blob.go [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+		return nil, ErrNotBlob
+	}
+
+	return &Blob{
+		TreeEntry: &TreeEntry{
+			mode: EntryBlob,
+			typ:  ObjectBlob,
+			id:   MustIDFromString(rev),
+			parent: &Tree{
+				repo: repo,
```
Comment (unknwon) [inline repo_blob.go:29]:
```suggestion
				repo: r,
```

**Comment by unknwon** [comment]
There are too many compile errors, could you fix that?

**Comment by jonatan5524** [comment]
> There are too many compile errors... could you fix that?

yes, I am not working from home, working from web, because of that there are errors, I will take care of it.

**Comment by jonatan5524** [comment]
fixed the compile errors @unknwon , sorry for the delay and the bad work, I had a busy week and could work on it properly only now.

**Comment by unknwon** [comment]
No worries at all! I’ll take another look.

**Comment by unknwon** [comment]
https://github.com/gogs/git-module/releases/tag/v1.7.0 has been tagged for this merge.

**Pull request merged**